numba/targets/arraymath.py

```diff
@@ -2867,6 +2868,26 @@ def np_imag_impl(a):
 #----------------------------------------------------------------------------
 # Misc functions
 
+@overload(np.count_nonzero)
+def np_count_nonzero(arr, axis=None):
+    if not type_can_asarray(arr):
+        raise TypingError("The argument to np.count_nonzero must be array-like")
+
+    if (NumpyVersion(np.__version__) < '1.12.0'):
+        raise TypingError("axis is not supported on Numpy versions < 1.12.0")
+
+    if (axis in (None, types.none)):
+        def impl(arr, axis=None):
+            arr2 = np.ravel(arr)
+            return np.sum(arr2 != 0)
+    else:
+        def impl(arr, axis=None):
+            arr2 = arr.astype(np.bool_)
+            return np.sum(arr2, axis=axis)
```

Comment:
pretty sure that this will fall over with optional types, but fixing this is out of scope for this PR.

Reply:
Like this:
```python
from numba import njit
import numpy as np

@njit
def foo(c):
    x = np.arange(10)
    ax = None
    if c > 2:
        ax = 1
    return np.count_nonzero(x, axis=ax)

foo(1)
```
issue is in `np.sum` with axis impl, not in your code.

Reply:
I know that @luk-f-a is working on the `np.sum` code to add new features, and in the process we realised that there's need to rewrite a lot of it. So as to avoid conflict/duplicated work  I think an issue reminding us of the problem above is likely best for now. Thanks.

Reply:
Will do that.